diff --git a/rild/rild.c b/rild/rild.c
index 041413d..c12fe90 100644
--- a/rild/rild.c
+++ b/rild/rild.c
@@ -126,7 +126,7 @@ int main(int argc, char **argv) {
     RLOGD("**RIL Daemon Started**");
     RLOGD("**RILd param count=%d**", argc);
 
-    initWithMmapSize();
+//    initWithMmapSize();
     umask(S_IRGRP | S_IWGRP | S_IXGRP | S_IROTH | S_IWOTH | S_IXOTH);
     for (i = 1; i < argc ;) {
         if (0 == strcmp(argv[i], "-l") && (argc - i > 1)) {
