diff --git a/common/file.te b/common/file.te
index 9c083aa..9474c83 100644
--- a/common/file.te
+++ b/common/file.te
@@ -1,5 +1,5 @@
 # Default type for anything under /firmware.
-type firmware_file, fs_type, contextmount_type;
+#type firmware_file, fs_type, contextmount_type, vendor_file_type;
 
 #Define the qmux socket type
 type qmuxd_socket, file_type;
@@ -12,7 +12,7 @@ type pps_socket, file_type;
 
 # Define cnd socket and data file type
 type cnd_socket, file_type, mlstrustedobject;
-type cnd_data_file, file_type, data_file_type;
+type cnd_data_file, file_type, data_file_type, core_data_file_type;
 type chre_socket, file_type;
 
 # Define dpmd data file type
@@ -28,7 +28,7 @@ type sysfs_smd_open_timeout, sysfs_type, fs_type;
 
 #Define the files written during the operation of netmgrd and qmuxd
 type netmgrd_data_file, file_type, data_file_type;
-type data_test_data_file, file_type, data_file_type;
+type data_test_data_file, file_type, data_file_type, core_data_file_type;
 type sysrq_trigger_proc, fs_type, mlstrustedobject;
 # Persist file types
 type persist_file, file_type;
@@ -36,22 +36,22 @@ type persist_bluetooth_file, file_type;
 type persist_data_file, file_type;
 type persist_drm_file, file_type;
 type data_qtee_file, file_type, data_file_type;
-type data_qsee_file, file_type, data_file_type;
+type data_qsee_file, file_type, data_file_type, core_data_file_type;
 type persist_misc_file, file_type;
 type persist_bms_file, file_type;
 type persist_secnvm_file, file_type;
 
-type diag_data_file, file_type, data_file_type;
+type diag_data_file, file_type, data_file_type, core_data_file_type;
 
 #file type for restricting proc read by audiod
-type proc_audiod, fs_type;
+type proc_audiod, fs_type ,proc_type;
 
 #file type for irqbalance socket
 type msm_irqbalance_socket, file_type;
 
 # Sensor file types
 type sensors_socket, file_type;
-type sensors_data_file, file_type, data_file_type;
+type sensors_data_file, file_type, data_file_type, core_data_file_type;
 type sensors_persist_file, file_type;
 type sysfs_sensors, sysfs_type, fs_type;
 
@@ -61,9 +61,9 @@ type thermal_socket, file_type;
 type sysfs_msmuart_file, sysfs_type, fs_type;
 
 # Storage RFS file types
-type rfs_file, file_type, data_file_type;
+type rfs_file, file_type, data_file_type, core_data_file_type;
 type rfs_system_file, file_type;
-type rfs_shared_hlos_file, file_type, data_file_type;
+type rfs_shared_hlos_file, file_type, data_file_type, core_data_file_type;
 
 #mm-pp-daemon file type for sysfs access
 #type sysfs_leds, fs_type, sysfs_type;
@@ -94,8 +94,8 @@ type sysfs_spmi_dev, sysfs_type, fs_type;
 type sysfs_mpdecision, fs_type, sysfs_type;
 type sysfs_rqstats, fs_type, sysfs_type;
 type sysfs_cpu_online, fs_type, sysfs_type;
-type mpctl_socket, file_type, data_file_type, mlstrustedobject;
-type mpctl_data_file, file_type, data_file_type;
+type mpctl_socket, file_type, data_file_type, mlstrustedobject, core_data_file_type;
+type mpctl_data_file, file_type, data_file_type, core_data_file_type;
 
 #Define the files used by lm
 type lm_data_file, file_type, data_file_type;
@@ -115,13 +115,13 @@ type gamed_socket, file_type;
 
 #define the files writter during the operatio of iop
 type iop_socket, file_type;
-type iop_data_file, file_type, data_file_type;
+type iop_data_file, file_type, data_file_type, core_data_file_type;
 
 # SPSS Apps images location
-type spss_data_file, file_type, data_file_type;
+type spss_data_file, file_type, data_file_type, core_data_file_type;
 
 #mm-qcamera-daemon socket
-type camera_socket, file_type, data_file_type;
+type camera_socket, file_type, data_file_type, core_data_file_type;
 
 #Socket node needed by ims_data daemon
 type ims_socket, file_type;
@@ -136,8 +136,8 @@ type ssgqmig_socket, file_type, mlstrustedobject;
 type ssgtzd_socket, file_type, mlstrustedobject;
 
 #location file types
-type location_data_file, file_type, data_file_type;
-type location_socket, file_type, data_file_type;
+type location_data_file, file_type, data_file_type, core_data_file_type;
+type location_socket, file_type, data_file_type, core_data_file_type;
 type location_app_data_file, file_type, data_file_type;
 
 #File types required by mdm-helper
@@ -154,7 +154,7 @@ type sysfs_socinfo, fs_type, sysfs_type;
 type sysfs_usb_uicc, sysfs_type, fs_type;
 
 type qlogd_socket, file_type, mlstrustedobject;
-type qlogd_data_file, file_type, data_file_type;
+type qlogd_data_file, file_type, data_file_type, core_data_file_type;
 #Defines the files (configs, dumps, etc) used by display processes
 type display_vendor_data_file, file_type, data_file_type;
 
@@ -163,33 +163,33 @@ type persist_display_file, file_type;
 
 # IPA file types
 type ipacm_socket, file_type;
-type ipacm_data_file, file_type, data_file_type;
+type ipacm_data_file, file_type, data_file_type, core_data_file_type;
 type ipa_vendor_data_file, file_type, data_file_type;
 
 # Port-bridge file types
-type port_bridge_data_file, file_type, data_file_type;
+type port_bridge_data_file, file_type, data_file_type, core_data_file_type;
 
-type fm_data_file, file_type, data_file_type;
+type fm_data_file, file_type, data_file_type, core_data_file_type;
 
 #Define the files written during the operation of mmi
-type mmi_data_file, file_type, data_file_type;
+type mmi_data_file, file_type, data_file_type, core_data_file_type;
 
 #bluetooth firmware file types
 type bt_firmware_file, fs_type, contextmount_type;
 
 #needed by vold
-type  proc_dirty_ratio, fs_type;
+type  proc_dirty_ratio, fs_type ,proc_type;
 
 #File types by mmi
 type mmi_socket, file_type;
 
 # hbtp config file
-type hbtp_cfg_file, file_type;
+type hbtp_cfg_file, file_type, vendor_file_type;
 type hbtp_log_file, file_type, data_file_type;
 type hbtp_kernel_sysfs, file_type, sysfs_type;
 
 #Define the files written during the operation of usf
-type usf_data_file, file_type, data_file_type;
+type usf_data_file, file_type, data_file_type, core_data_file_type;
 type persist_usf_file, file_type;
 
 #qfp-daemon
@@ -197,7 +197,7 @@ type qfp-daemon_data_file, file_type, data_file_type;
 type persist_qti_fp_file, file_type;
 
 # dts notifier files
-type dts_data_file, file_type, data_file_type;
+type dts_data_file, file_type, data_file_type, core_data_file_type;
 
 #qsee_svc_app file types
 type qsee_svc_app_data_file, file_type, data_file_type;
@@ -219,22 +219,22 @@ type qti_logkit_pub_socket, file_type, mlstrustedobject, data_file_type;
 type adsprpcd_file, file_type, mlstrustedobject;
 
 # audio pp notifier files
-type audio_pp_data_file, file_type, data_file_type;
+type audio_pp_data_file, file_type, data_file_type, core_data_file_type;
 
 #mdtp_svc_app file types
 type mdtp_svc_app_data_file, file_type, data_file_type;
 
 # subsystem_ramdump files
-type ssr_ramdump_data_file, file_type, data_file_type;
+type ssr_ramdump_data_file, file_type, data_file_type, core_data_file_type;
 
 # Regionalization files
-type regionalization_file, file_type;
+type regionalization_file, file_type, vendor_file_type;
 
 # /data/system/swap/swapfile - swapfile
-type swap_data_file, file_type, data_file_type;
+type swap_data_file, file_type, data_file_type, core_data_file_type;
 
 # dynamic nv files
-type dynamic_nv_data_file, file_type, data_file_type;
+type dynamic_nv_data_file, file_type, data_file_type, core_data_file_type;
 
 # Wifi Data file
 type wifi_vendor_data_file, file_type, data_file_type;
@@ -314,7 +314,7 @@ type sysfs_msm_stats, fs_type, sysfs_type;
 type sysfs_msm_power, fs_type, sysfs_type;
 
 # Data type for QVOP
-type qvop-daemon_data_file, file_type, data_file_type;
+type qvop-daemon_data_file, file_type, data_file_type, core_data_file_type;
 
 type sysfs_fm, sysfs_type, fs_type;
 
diff --git a/common/rfs_access.te b/common/rfs_access.te
index 46b224d..37f8b57 100644
--- a/common/rfs_access.te
+++ b/common/rfs_access.te
@@ -70,5 +70,7 @@ allow rfs_access self:capability { dac_read_search chown dac_override };
 allow rfs_access kmsg_device:chr_file w_file_perms;
 
 #Prevent other domains from accessing RFS data files.
-neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell } rfs_file:dir create_dir_perms;
-neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell } rfs_file:file create_file_perms;
+neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell } rfs_file:dir {reparent rename lock lock };
+neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell -vendor_init } rfs_file:dir { ioctl read write create getattr setattr add_name remove_name search rmdir open };
+neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell } rfs_file:file { rename ioctl lock map append lock map };
+neverallow { domain -rfs_access -kernel -recovery -init userdebug_or_eng(`-su') -qti_init_shell -vendor_init } rfs_file:file { read write create getattr setattr unlink open };
diff --git a/legacy-common/bluetooth_loader.te b/legacy-common/bluetooth_loader.te
index b969bdc..a86c39b 100644
--- a/legacy-common/bluetooth_loader.te
+++ b/legacy-common/bluetooth_loader.te
@@ -1,5 +1,5 @@
 type bluetooth_loader, domain;
-type bluetooth_loader_exec, exec_type, file_type;
+type bluetooth_loader_exec, exec_type, file_type, vendor_file_type;
 
 # Started by init
 init_daemon_domain(bluetooth_loader)
